From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Max Filippov <jcmvbkbc@gmail.com>
Date: Tue, 10 Oct 2023 17:02:39 -0700
Subject: [PATCH 21/61] drivers/net/wireless/espressif/esp32-ng: update

Update to upstream commit d07a9242f4905f4423ef5289af1a2fbafc0388c6

Signed-off-by: Max Filippov <jcmvbkbc@gmail.com>
(cherry picked from commit 4b5bf822dd4d15046b342665b6719c144e1b1ce2)
---
 drivers/net/wireless/espressif/esp32-ng/adapter.h   |  2 ++
 .../net/wireless/espressif/esp32-ng/esp_cfg80211.c  |  5 +++++
 drivers/net/wireless/espressif/esp32-ng/esp_cmd.c   | 13 ++++++++++++-
 .../wireless/espressif/esp32-ng/esp_kernel_port.h   |  6 +++---
 drivers/net/wireless/espressif/esp32-ng/esp_sdio.c  |  2 +-
 .../net/wireless/espressif/esp32-ng/esp_sdio_api.c  |  2 +-
 .../net/wireless/espressif/esp32-ng/esp_sdio_api.h  |  2 +-
 .../net/wireless/espressif/esp32-ng/esp_sdio_decl.h |  2 +-
 drivers/net/wireless/espressif/esp32-ng/esp_spi.c   |  2 +-
 drivers/net/wireless/espressif/esp32-ng/esp_spi.h   |  2 +-
 10 files changed, 28 insertions(+), 10 deletions(-)

diff --git a/drivers/net/wireless/espressif/esp32-ng/adapter.h b/drivers/net/wireless/espressif/esp32-ng/adapter.h
index 9fd8a7742f826..0efe5ddac3595 100644
--- a/drivers/net/wireless/espressif/esp32-ng/adapter.h
+++ b/drivers/net/wireless/espressif/esp32-ng/adapter.h
@@ -170,6 +170,8 @@ struct cmd_sta_auth {
 	uint8_t    channel;
 	uint8_t    auth_type;
 	char       ssid[MAX_SSID_LEN+1];
+	uint8_t    key_len;
+	uint8_t    key[27];
 	uint8_t    auth_data_len;
 	uint8_t    pad[2];
 	uint8_t    auth_data[];
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.c b/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.c
index 935189ef44f90..5273d560995f0 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.c
@@ -607,6 +607,7 @@ static void esp_reg_notifier(struct wiphy *wiphy,
 {
 	struct esp_wifi_device *priv = NULL;
 	struct esp_device *esp_dev = NULL;
+	struct esp_adapter *adapter = esp_get_adapter();
 
 	if (!wiphy || !request) {
 		esp_info("%u invalid input\n", __LINE__);
@@ -619,6 +620,10 @@ static void esp_reg_notifier(struct wiphy *wiphy,
 		esp_info("%u esp_dev not initialized yet \n", __LINE__);
 		return;
 	}
+	if (test_bit(ESP_CLEANUP_IN_PROGRESS, &adapter->state_flags)) {
+               return;
+	}
+
 	priv = esp_dev->adapter->priv[0];
 
 	if (!priv) {
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_cmd.c b/drivers/net/wireless/espressif/esp32-ng/esp_cmd.c
index ca51198f6635b..29bc3c99e64d4 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_cmd.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_cmd.c
@@ -495,6 +495,7 @@ static void process_scan_result_event(struct esp_wifi_device *priv,
 	u16 cap_info;
 	u32 ie_len;
 	int freq;
+	int frame_type = CFG80211_BSS_FTYPE_UNKNOWN; /* int type for older compatibilty */
 
 	if (!priv || !scan_evt) {
 		esp_err("Invalid arguments\n");
@@ -531,9 +532,15 @@ static void process_scan_result_event(struct esp_wifi_device *priv,
 	ie_buf += sizeof(struct beacon_probe_fixed_params);
 	ie_len -= sizeof(struct beacon_probe_fixed_params);
 
+	if ((scan_evt->frame_type << 4) == IEEE80211_STYPE_BEACON) {
+		frame_type = CFG80211_BSS_FTYPE_BEACON;
+	} else if ((scan_evt->frame_type << 4) == IEEE80211_STYPE_PROBE_RESP) {
+		frame_type = CFG80211_BSS_FTYPE_PRESP;
+	}
+
 	if (chan && !(chan->flags & IEEE80211_CHAN_DISABLED)) {
 		bss = CFG80211_INFORM_BSS(priv->adapter->wiphy, chan,
-				scan_evt->bssid, timestamp,
+				frame_type, scan_evt->bssid, timestamp,
 				cap_info, beacon_interval, ie_buf, ie_len,
 				(le32_to_cpu(scan_evt->rssi) * 100), GFP_ATOMIC);
 
@@ -944,6 +951,10 @@ int cmd_auth_request(struct esp_wifi_device *priv,
 	cmd->auth_data_len = req->auth_data_len;
 	memcpy(cmd->auth_data, req->auth_data, req->auth_data_len);
 
+	if (req->key_len) {
+		memcpy(cmd->key, req->key, req->key_len);
+		cmd->key_len = req->key_len;
+	}
 	esp_info("Authentication request: "MACSTR" %d %d %d %d\n",
 			MAC2STR(cmd->bssid), cmd->channel, cmd->auth_type, cmd->auth_data_len,
 			(u32) req->ie_len);
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_kernel_port.h b/drivers/net/wireless/espressif/esp32-ng/esp_kernel_port.h
index f7cfd3fae3504..03d11773e54ba 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_kernel_port.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_kernel_port.h
@@ -56,14 +56,14 @@ static inline void ether_addr_copy(u8 *dst, const u8 *src)
 
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3, 18, 0)
-  #define CFG80211_INFORM_BSS(wiphy, chan, bssid, tsf, \
+  #define CFG80211_INFORM_BSS(wiphy, chan, type, bssid, tsf, \
 	  cap, beacon_interval, ie, ielen, sig, gfp) \
   cfg80211_inform_bss(wiphy, chan, bssid, tsf, \
 	  cap, beacon_interval, ie, ielen, sig, gfp)
 #else
-  #define CFG80211_INFORM_BSS(wiphy, chan, bssid, tsf, \
+  #define CFG80211_INFORM_BSS(wiphy, chan, type, bssid, tsf, \
 	  cap, beacon_interval, ie, ielen, signal, gfp) \
-  cfg80211_inform_bss(wiphy, chan, CFG80211_BSS_FTYPE_UNKNOWN, bssid, tsf, \
+  cfg80211_inform_bss(wiphy, chan, type, bssid, tsf, \
 	  cap, beacon_interval, ie, ielen, signal, gfp)
 #endif
 
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_sdio.c b/drivers/net/wireless/espressif/esp32-ng/esp_sdio.c
index a8ab014891ad1..c986ecf7927f1 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_sdio.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_sdio.c
@@ -1,9 +1,9 @@
+// SPDX-License-Identifier: GPL-2.0-only
 /*
  * Espressif Systems Wireless LAN device driver
  *
  * SPDX-FileCopyrightText: 2015-2023 Espressif Systems (Shanghai) CO LTD
  *
- * SPDX-License-Identifier: GPL-2.0-only
  */
 #include "utils.h"
 #include <linux/mutex.h>
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_sdio_api.c b/drivers/net/wireless/espressif/esp32-ng/esp_sdio_api.c
index 49eb573433fb5..889ee8ffec8ac 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_sdio_api.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_sdio_api.c
@@ -1,9 +1,9 @@
+// SPDX-License-Identifier: GPL-2.0-only
 /*
  * Espressif Systems Wireless LAN device driver
  *
  * SPDX-FileCopyrightText: 2015-2023 Espressif Systems (Shanghai) CO LTD
  *
- * SPDX-License-Identifier: GPL-2.0-only
  */
 #include "utils.h"
 #include <linux/mmc/sdio.h>
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_sdio_api.h b/drivers/net/wireless/espressif/esp32-ng/esp_sdio_api.h
index 718395f3d8c80..a07892442dca8 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_sdio_api.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_sdio_api.h
@@ -1,9 +1,9 @@
+// SPDX-License-Identifier: GPL-2.0-only
 /*
  * Espressif Systems Wireless LAN device driver
  *
  * SPDX-FileCopyrightText: 2015-2023 Espressif Systems (Shanghai) CO LTD
  *
- * SPDX-License-Identifier: GPL-2.0-only
  */
 #ifndef _ESP_SDIO_API_H_
 #define _ESP_SDIO_API_H_
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_sdio_decl.h b/drivers/net/wireless/espressif/esp32-ng/esp_sdio_decl.h
index 1a4088ad1daf8..1d2298917f0d5 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_sdio_decl.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_sdio_decl.h
@@ -1,9 +1,9 @@
+// SPDX-License-Identifier: GPL-2.0-only
 /*
  * Espressif Systems Wireless LAN device driver
  *
  * SPDX-FileCopyrightText: 2015-2023 Espressif Systems (Shanghai) CO LTD
  *
- * SPDX-License-Identifier: GPL-2.0-only
  */
 #ifndef _ESP_DECL_H_
 #define _ESP_DECL_H_
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_spi.c b/drivers/net/wireless/espressif/esp32-ng/esp_spi.c
index 14997d00f4013..aca8b6757e111 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_spi.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_spi.c
@@ -1,7 +1,7 @@
+// SPDX-License-Identifier: GPL-2.0-only
 /*
  * SPDX-FileCopyrightText: 2015-2023 Espressif Systems (Shanghai) CO LTD
  *
- * SPDX-License-Identifier: GPL-2.0-only
  */
 #include "utils.h"
 #include <linux/spi/spi.h>
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_spi.h b/drivers/net/wireless/espressif/esp32-ng/esp_spi.h
index 00d53e13b62e7..3d0666aede290 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_spi.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_spi.h
@@ -1,7 +1,7 @@
+// SPDX-License-Identifier: GPL-2.0-only
 /*
  * SPDX-FileCopyrightText: 2015-2023 Espressif Systems (Shanghai) CO LTD
  *
- * SPDX-License-Identifier: GPL-2.0-only
  */
 
 #ifndef _ESP_SPI_H_
-- 
2.55.0

